python/src/aac/parser.py

```diff
@@ -8,11 +8,105 @@
+    column: int = attrib(validator=validators.instance_of(int))
+    position: int = attrib(validator=validators.instance_of(int))
+    span: int = attrib(validator=validators.instance_of(int))
+
+
+@attrs
+class Lexeme:
+    """A lexical unit for a parsed AaC object.
+
+    Attributes:
+        location (SourceLocation): The location at which the object was found.
+        source (str): The source in which the object was found.
+        value (str): The value of the parsed object.
+    """
+
+    location: SourceLocation = attrib(validator=validators.instance_of(SourceLocation))
+    source: str = attrib(validator=validators.instance_of(str))
+    value: str = attrib(validator=validators.instance_of(str))
+
+
+@attrs
+class ParsedModel:
+    """A parsed Architecture-as-Code model.
+
+    Attributes:
+        content (str): The textual representation of the model.
+        lexemes (list[Lexeme]): A list of lexemes for each item in the parsed model.
+        model (dict): The parsed model.
+    """
+
+    content: str = attrib(validator=validators.instance_of(str))
+    lexemes: list[Lexeme] = attrib(default=Factory(list), validator=validators.instance_of(list))
+    model: dict = attrib(default=Factory(list), validator=validators.instance_of(dict))
+
+
+def parse(source: str) -> ParsedModel:
+    """Parse the Architecture-as-Code (AaC) model(s) from the provided source.
+
+    Args:
+        source (str): The location from which to extract the Architecture-as-Code content.
+
+    Returns:
+        A ParsedModel object containing the internal representation of the model and metadata
+        associated with the model.
+    """
+
+    def mark_to_source_location(start: yaml.error.Mark, end: yaml.error.Mark) -> SourceLocation:
+        return SourceLocation(start.line, start.column, start.index, end.column - start.column)
+
+    def get_lexemes_for_model(contents):
+        yaml_source = os.path.abspath(source) if is_path_name(source) else "<string>"
+        lexemes = []
+        tokens = [token for token in yaml.scan(contents, Loader=yaml.SafeLoader)]
+        for token in tokens:
+            if hasattr(token, "value"):
+                location = mark_to_source_location(token.start_mark, token.end_mark)
+                lexemes.append(Lexeme(location, yaml_source, token.value))
+        return lexemes
+
+    contents = get_yaml_from_source(source)
+    model = _parse_file(source) if is_path_name(source) else _parse_str(source, contents)
+    return ParsedModel(contents, get_lexemes_for_model(contents), model)
+
+
+def get_yaml_from_source(source: str) -> str:
+    """Get the YAML contents from the provided source.
+
+    Args:
+        source (str): The source from which to get the YAML contents. This can be a path-like
+        string pointing to a file with YAML contents; or a string of YAML contents.
+
+    Returns:
+        The YAML contents extracted from source.
+    """
+    if is_path_name(source):
```

Comment:
I think this makes it more explicit to the reader what is being tested. If is_path_name were more complex, then I'd agree with the abstraction. If `os.path.lexists` is too verbose `path.lexists` should work too.
```suggestion
    if os.path.lexists(source):
```

Reply:
Here's another thought: This change doesn't really allow us to give feedback if a user entered an incorrect file path. 

This check is assuming that if it's not a file that exists on the filesystem then it's a yaml string, which may not be the case. I think we could circumvent this issue if we did some argument validation, like plugins that list command arguments as being a `file` type would assert that those file arguments are in fact files that exist on the filesystem or kick back an error before even running the plugin command.

Reply:
I agree that we should do this. Perhaps we can create another story to handle this more thorough style of validation?

Reply:
Yeah, that would certainly be a separate story.